Mathlib/Algebra/Category/GroupCat/Injective.lean

```diff
@@ -43,384 +44,129 @@ set_option linter.uppercaseLean3 false
-    LinearEquiv.coe_toEquiv]
-  rw [← Submodule.Quotient.equiv_refl, Submodule.Quotient.equiv_symm,
-    Submodule.Quotient.equiv_apply]
-  convert Submodule.liftQSpanSingleton_apply _ _ _ _ using 1
-  · convert (LinearMap.toSpanSingleton_ker a).symm using 1
-    exact Submodule.map_id _
-
-lemma toRatCircle_apply_self_eq :
-    toRatCircle infinite_order ⟨a, Submodule.mem_span_singleton_self a⟩ =
-    Quotient.mk _ (2 : ℚ)⁻¹ := by
-  rw [← toRatCircle_apply_self_eq_aux infinite_order]
-  simp
-
-lemma toRatCircle_apply_self_ne_zero :
-    toRatCircle infinite_order ⟨a, Submodule.mem_span_singleton_self a⟩ ≠ 0 := by
-  rw [toRatCircle_apply_self_eq infinite_order]
-  change _ ≠ Quotient.mk _ (0 : ℚ)
-  intro r
-  erw [QuotientAddGroup.mk'_eq_mk'] at r
-  obtain ⟨_, ⟨z, rfl⟩, H⟩ := r
-  simp only [zsmul_eq_mul, mul_one] at H
-  rw [add_comm, add_eq_zero_iff_eq_neg] at H
-  have ineq0 : |(z : ℚ)| < 1
-  · rw [H, abs_neg, abs_inv, inv_lt_one_iff]; right; norm_num
-  norm_cast at ineq0
-  rw [Int.abs_lt_one_iff] at ineq0
-  subst ineq0
-  norm_num at H
-
-end infinite_order
-
-namespace finite_order
-
-variable {A_}
-variable {a : A_} (finite_order : IsOfFinAddOrder a)
-
-lemma divBy_addOrderOf_addOrderOf : divBy (addOrderOf a) (addOrderOf a) = 0 := by
-  simp only [divBy_apply, Int.cast_ofNat, ne_eq, Nat.cast_eq_zero]
-  apply Quotient.sound'
-  rw [QuotientAddGroup.leftRel_eq]
-  simp only [ne_eq, Nat.cast_eq_zero, add_zero, neg_mem_iff]
-  rw [mul_inv_cancel]
-  · exact AddSubgroup.mem_zmultiples 1
-  · rw [← addOrderOf_pos_iff] at finite_order
-    norm_num
-    linarith
-
-/-- the map sending `n • a` to `n / ord(a)` when `a` has finite order. -/
+  rw [Ideal.mem_span_singleton, addOrderOf_dvd_iff_zsmul_eq_zero]
+  rfl
+
+/-- `ℤ ⧸ ⟨ord(a)⟩ ≃ aℤ` -/
+@[simps!] noncomputable def equivZModSpanAddOrderOf :
+    (ℤ ∙ a) ≃ₗ[ℤ] ℤ ⧸ Ideal.span {(addOrderOf a : ℤ)} :=
+  (LinearEquiv.ofEq _ _ <| LinearMap.span_singleton_eq_range ℤ A_ a).trans <|
+    (LinearMap.quotKerEquivRange <| LinearMap.toSpanSingleton ℤ A_ a).symm.trans <|
+      Submodule.quotEquivOfEq _ _ <| LinearMap.toSpanSingleton_ker a
+
+lemma equivZModSpanAddOrderOf_apply_self :
+    equivZModSpanAddOrderOf a ⟨a, Submodule.mem_span_singleton_self a⟩ = Submodule.Quotient.mk 1 :=
+  (LinearEquiv.eq_symm_apply _).mp (one_zsmul _).symm
+
+/-- Given `n : ℕ`, the map `m ↦ m / n`. -/
+abbrev divBy (n : ℕ) : ℤ →ₗ[ℤ] AddCircle (1 : ℚ) :=
+  LinearMap.toSpanSingleton ℤ _ (QuotientAddGroup.mk (n : ℚ)⁻¹)
+
+lemma divBy_self (n : ℕ) : divBy n n = 0 := by
+  obtain rfl | h0 := eq_or_ne n 0
+  · apply map_zero
+  apply (AddCircle.coe_eq_zero_iff _).mpr ⟨1, _⟩
+  simp [mul_inv_cancel (Nat.cast_ne_zero (R := ℚ).mpr h0)]
+
+variable {a}
+
+/-- The map sending `n • a` to `n / 2` when `a` has infinite order,
+  and to `n / addOrderOf a` otherwise. -/
 @[simps!] noncomputable def toRatCircle : (ℤ ∙ a) →ₗ[ℤ] AddCircle (1 : ℚ) :=
```

Comment:
Should this actually be renamed to `toAddCircle`?

Reply:
RatCircle makes sense as it distinguishes Q/Z from the usual circle R/Z. Anyway this declaration is in a namespace not supposed to be used from outside, so the name shouldn't matter much.